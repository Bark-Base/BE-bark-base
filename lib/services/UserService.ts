import * as bcrypt  from 'bcrypt';
const jwt = require('jsonwebtoken');
const User = require('../models/User');

module.exports = class UserService {
  static async create({ email, password, secretAnswer }: {email:string, password:string, secretAnswer:string}) {
    const passwordHash = await bcrypt.hash(
      password, 
      Number(process.env.SALT_ROUNDS)

    );
    const secretHash = await bcrypt.hash(
        secretAnswer, 
        Number(process.env.SALT_ROUNDS)
      );

    const user = await User.insert({
      email,
      passwordHash,
      secretHash,
    });
    return user;
  }

  static async signIn({ email, password = '' } :{email:string, password:string }) {
    try {
      const user = await User.getByEmail(email);

      if(!user) throw new Error('Invalid email');
      if(!bcrypt.compareSync(password, user.passwordHash))
        throw new Error('Invalid password');
        
      const token = jwt.sign({ ...user }, process.env.JWT_SECRET, {
        expiresIn: '1 day',
      });
      return token;
    } catch (error:any) {
      error.status = 401;
      throw error;
    }
  }
};