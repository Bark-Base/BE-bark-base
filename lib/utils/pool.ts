import { Pool } from 'pg';

export  const pool = new Pool({
  connectionString: process.env.DATABASE_URL || process.env.LOCAL_DATABASE_URL,
  ssl:  process.env.DATABASE_URL ? true : false  && { rejectUnauthorized : false }
});

// eslint-disable-next-line no-console
pool.on('connect', () => console.log('🐘 Postgres connected'));
