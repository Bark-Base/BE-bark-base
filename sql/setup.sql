DROP TABLE IF EXISTS users Cascade;
DROP TABLE IF EXISTS pets Cascade;
DROP TABLE IF EXISTS contacts Cascade;
DROP TABLE IF EXISTS medical_info Cascade;

CREATE TABLE users (
    owner_id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    password_hash TEXT NOT NULL,
    secret_hash TEXT,
    email TEXT NOT NULL UNIQUE
);
INSERT INTO users (password_hash, secret_hash, email)
VALUES('$2b$10$cYidLM7q1F449Qm3a/a9a.XbqC.U24OQCcYkVyYZTwGkIhv8cFIk6', 'shhdhshshsh', 'barkbase@web.com');

CREATE TABLE pets (
    pet_id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    owner_id BIGINT REFERENCES users,
    name TEXT NOT NULL,
    birthday VARCHAR(10),
    image_url TEXT
);
INSERT INTO pets (owner_id, name, birthday, image_url)
VALUES(1,'Kharkiv','2022-01-01','http://www.placekitten.com');
INSERT INTO pets (owner_id, name, birthday, image_url)
VALUES(1,'Kyiv','2022-01-01','http://www.placekitten.com');
INSERT INTO pets (owner_id, name, birthday, image_url)
VALUES(1,'Odessa','2022-01-01','http://www.placekitten.com');

CREATE TABLE contacts (
    contact_id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    type TEXT NOT NULL,
    name TEXT NOT NULL,
    phone VARCHAR(20),
    email TEXT,
    address TEXT,
    owner_id BIGINT REFERENCES users,
    pet_id BIGINT REFERENCES pets ON DELETE CASCADE
);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('vet','bob','555-555-1234','barkbase@gmail.com','123 bake st.',1,2);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('trainer','fred','111-555-5555','sample@gmail.com','456 cake st.',1,2);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('walker','texas','555-867-5309','ranger@gmail.com','789 faque st.',1,2);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('vet','bobbina','555-545-5555','ample@gmail.com','123 fake st.',1,1);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('trainer','frederick','555-555-5555','sample@gmail.com','123 fake st.',1,1);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('walker','etcetera','555-555-5555','sample@gmail.com','1213 fake st.',1,1);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('vet','bob ross','555-555-4656','simple@gmail.com','4123 lake st.',1,3);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('trainer','Jamba Juice','555-555-5555','sample@gmail.com','5123 pake st.',1,3);
INSERT INTO contacts (type,name,phone,email,address,owner_id,pet_id)
VALUES('walker','Todd','555-555-5555','sample@gmail.com','6123 make st.',1,3);

CREATE TABLE medical_info (
    medical_id BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    vet_id BIGINT REFERENCES contacts(contact_id),
    medicines VARCHAR(30), 
    notes VARCHAR(2000),
    next_appt DATE,
    pet_id BIGINT REFERENCES pets ON DELETE CASCADE
);

INSERT INTO medical_info (vet_id, medicines, notes, next_appt, pet_id)
VALUES(1,'scotch','Take twice hourly', '2022-02-02', 1 );