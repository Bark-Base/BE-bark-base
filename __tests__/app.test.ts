import { pool } from '../lib/utils/pool';
import { setup } from '../data/setup';
import  request  from 'supertest';
import UserService from '../lib/services/UserService';
const app = require('../lib/app');
const mockUser = { email: 'test@user', password: '123' };
const mockUser2 = { email: 'test@user2', password: '123' };
const registerAndLogin = async (userProps:any = {}) => {
  const password = userProps.password ?? mockUser.password;
  
  // agent dependency allows the storage of cookies for tests
  const agent:any = request.agent(app);
  
  // Call create user method
  const user = await UserService.create({ ...mockUser, ...userProps });
  
  // after user creation this signs into the page
  const { email } = user;
  await agent.post('/api/v1/users/sessions').send({ email, password });
  return [agent, user];
};

describe('user auth routes',  () => {
  beforeAll(() => {
    return setup(pool);
  });

  afterAll(() => {
    pool.end();
  });

  it('tests the signup route', async () => {
    const res = await request(app).post('/api/v1/auth').send(mockUser2);
    expect(res.body).toEqual({
        
      message: 'Signed in successfully!',
      user: {
        ownerId: expect.any(String),
        email: 'test@user2'
      }
        
    });
    
  });
  it('should return a 401 when signed out and trying to view user', async () => {
    const res = await request(app).get('/api/v1/auth/user');
    expect(res.text).toMatchSnapshot();
  });

  // it('should create a pet', async () => {
    
  //   const [ user] = await registerAndLogin({ email:'test@user'});
  //   const check:any = await request(app).post('/api/v1/pet/').send({ownerId:'2', name:'bubba', birthday:'123', imageUrl:''}); 
    
  //   console.log(user, check.body);
  //   const res = await request(app).get(`/api/v1/pet/all/2`)

  //   expect(res.body).toEqual([{ ownerId: expect.any(String), }]);
  // });

});
