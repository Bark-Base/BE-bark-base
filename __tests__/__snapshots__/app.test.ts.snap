// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user auth routes should return a 401 when signed out and trying to view user 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
<meta charset=\\"utf-8\\">
<title>Error</title>
</head>
<body>
<pre>Error: You must be signed in to continue<br> &nbsp; &nbsp;at C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\lib\\\\middleware\\\\authenticate.ts:9:24<br> &nbsp; &nbsp;at Generator.next (&lt;anonymous&gt;)<br> &nbsp; &nbsp;at C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\lib\\\\middleware\\\\authenticate.ts:8:71<br> &nbsp; &nbsp;at new Promise (&lt;anonymous&gt;)<br> &nbsp; &nbsp;at Object.&lt;anonymous&gt;.__awaiter (C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\lib\\\\middleware\\\\authenticate.ts:4:12)<br> &nbsp; &nbsp;at Object.&lt;anonymous&gt;.exports.default (C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\lib\\\\middleware\\\\authenticate.ts:4:67)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\node_modules\\\\express\\\\lib\\\\router\\\\layer.js:95:5)<br> &nbsp; &nbsp;at next (C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\node_modules\\\\express\\\\lib\\\\router\\\\route.js:137:13)<br> &nbsp; &nbsp;at Route.dispatch (C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\node_modules\\\\express\\\\lib\\\\router\\\\route.js:112:3)<br> &nbsp; &nbsp;at Layer.handle [as handle_request] (C:\\\\Users\\\\14055\\\\alchemy\\\\be-bark-base\\\\node_modules\\\\express\\\\lib\\\\router\\\\layer.js:95:5)</pre>
</body>
</html>
"
`;
